typescript/sdk/src/ism/CosmosNativeIsmModule.ts

```diff
@@ -0,0 +1,163 @@
+import { SigningHyperlaneModuleClient } from '@hyperlane-xyz/cosmos-sdk';
+import {
+  Address,
+  ChainId,
+  Domain,
+  ProtocolType,
+  assert,
+  deepEquals,
+  rootLogger,
+} from '@hyperlane-xyz/utils';
+
+import {
+  HyperlaneModule,
+  HyperlaneModuleParams,
+} from '../core/AbstractHyperlaneModule.js';
+import { MultiProvider } from '../providers/MultiProvider.js';
+import { AnnotatedCosmJsNativeTransaction } from '../providers/ProviderType.js';
+import { ChainName, ChainNameOrId } from '../types.js';
+import { normalizeConfig } from '../utils/ism.js';
+
+import { CosmosNativeIsmReader } from './CosmosNativeIsmReader.js';
+import { IsmConfig, IsmConfigSchema, IsmType } from './types.js';
+
+type IsmModuleAddresses = {
+  deployedIsm: Address;
+  mailbox: Address;
+};
+
+export class CosmosNativeIsmModule extends HyperlaneModule<
+  ProtocolType.CosmosNative,
+  IsmConfig,
+  IsmModuleAddresses
+> {
+  protected readonly logger = rootLogger.child({
+    module: 'CosmosNativeIsmModule',
+  });
+  protected readonly reader: CosmosNativeIsmReader;
+  protected readonly mailbox: Address;
+
+  // Adding these to reduce how often we need to grab from MultiProvider.
+  public readonly chain: ChainName;
+  public readonly chainId: ChainId;
+  public readonly domainId: Domain;
+
+  constructor(
+    protected readonly multiProvider: MultiProvider,
+    params: HyperlaneModuleParams<IsmConfig, IsmModuleAddresses>,
+    protected readonly signer: SigningHyperlaneModuleClient,
+  ) {
+    params.config = IsmConfigSchema.parse(params.config);
+    super(params);
+
+    this.mailbox = params.addresses.mailbox;
+    this.chain = multiProvider.getChainName(this.args.chain);
+    this.chainId = multiProvider.getChainId(this.chain);
+    this.domainId = multiProvider.getDomainId(this.chain);
+
+    this.reader = new CosmosNativeIsmReader(signer);
+  }
+
+  public async read(): Promise<IsmConfig> {
+    return this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
+  }
+
+  // whoever calls update() needs to ensure that targetConfig has a valid owner
+  public async update(
+    expectedConfig: IsmConfig,
+  ): Promise<AnnotatedCosmJsNativeTransaction[]> {
+    expectedConfig = IsmConfigSchema.parse(expectedConfig);
+
+    // Do not support updating to a custom ISM address
+    if (typeof expectedConfig === 'string') {
+      throw new Error(
+        'Invalid targetConfig: Updating to a custom ISM address is not supported. Please provide a valid ISM configuration.',
+      );
+    }
+
+    // save current config for comparison
+    // normalize the config to ensure it's in a consistent format for comparison
+    const actualConfig = normalizeConfig(await this.read());
+    expectedConfig = normalizeConfig(expectedConfig);
+
+    assert(
+      typeof expectedConfig === 'object',
```

Comment:
is this needed? `if (typeof expectedConfig === 'string') { throw ... }` should ensure that it's not a string

Reply:
so it is also in `EvmIsmModule`: https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/troy/cosmos-ism-module/typescript/sdk/src/ism/EvmIsmModule.ts#L102, so I would also include it here